### PR TITLE
Exporter enhancements and refactor

### PR DESCRIPTION
**Pull request by clinta**
This is a significant refactor of the farcaster exporter.

I started with the goal of adding features to serve the metrics directly over http, rather than relying on the prometheus node exporter. This was challenging because of the global state and calls to terminate that caused the entire program to exit.

This refactor removes all global variables, implements separate daemon and miner subclasses of the lotus class, and uses context and `with as` statements to handle writing out the final metrics for collection time and success status.

As for features this adds several. First of all farcaster will now use the same environment variables that lotus does `FULLNODE_API_INFO` and `MINER_API_INFO` to discover the urls and tokens. Falling back to looking for the files in `LOTUS_PATH` and `LOTUS_MINER_PATH`. Command line arguments are also supported which take priority.

Farcaster can now write to a file directly without relying on redirection in a cron job, using the `--file` option. And the feature I originally set out to build, farcaster can now listen on a TCP port and serve metrics directly to prometheus for scraping. It also supports sytemd socket activation so that farcaster can automatically be started when prometheus tries to scrape it. Sample service and socket unit files have been added.

These changes should be backwards compatible and not change anything about the metrics exported.

### CODE REVIEW

**Comment by s0nik42** [comment]
Hi @clinta thank you very much for your PR !  let me look and it during the week.

File: lotus-exporter-farcaster/lotus-exporter-farcaster.py [hunk 1]
```diff
@@ -939,175 +850,157 @@ def load_toml(toml_file):
     try:
         with open(toml_file) as data_file:
             nested_dict = toml.load(data_file)
-    except Exception as e_generic:
-        terminate(f'Error: loading file {toml_file} : {str(e_generic)}')
+    except Exception as exp:
```
Comment (Breigner01) [inline lotus-exporter-farcaster/lotus-exporter-farcaster.py:853]: Do not except bare Exception it is too broad

File: lotus-exporter-farcaster/lotus-exporter-farcaster.py [hunk 2]
```diff
@@ -939,175 +850,157 @@ def load_toml(toml_file):
...
+    miner_worker_addr = daemon.get("StateAccountKey", [miner_worker, daemon.tipset_key()])["result"]
 
     # Add miner addresses to known_addresses lookup table
-    LOTUS_OBJ.add_known_addresses({miner_owner: "Local Owner", miner_owner_addr: "Local Owner", miner_worker: "Local Worker", miner_worker_addr: "Local Worker"})
+    daemon.add_known_addresses({miner_owner: "Local Owner", miner_owner_addr: "Local Owner", miner_worker: "Local Worker", miner_worker_addr: "Local Worker"})
 
     try:
         miner_control0 = daemon_stats["result"]["ControlAddresses"][0]
-    except Exception as e_generic:
+    except Exception:
```
Comment (Breigner01) [inline lotus-exporter-farcaster/lotus-exporter-farcaster.py:902]: Do not except bare Exception

File: lotus-exporter-farcaster/lotus-exporter-farcaster.py [hunk 3]
```diff
@@ -1133,48 +1026,48 @@ def main():
...
+    workerjobs = miner.get("WorkerJobs", [])
     for (wrk, job_list) in workerjobs["result"].items():
         for job in job_list:
             job_id = job['ID']['ID']
             sector = str(job['Sector']['Number'])
 
             try:
                 worker_host = workerstats["result"][wrk]["Info"]["Hostname"]
-            except Exception as e_generic:
+            except Exception:
```
Comment (Breigner01) [inline lotus-exporter-farcaster/lotus-exporter-farcaster.py:1049]: Do not except bare Exception

File: lotus-exporter-farcaster/lotus-exporter-farcaster.py [hunk 4]
```diff
@@ -1292,43 +1185,119 @@ def main():
...
+    with open(path.joinpath("api"), "r") as f:
+        [_, _, addr, _, port, proto] = f.read().split("/", 5)
+    url = f"{proto}://{addr}:{port}/rpc/v0"
+    return (url, token)
+
+def run(args, output):
+    with Metrics(output=output) as metrics:
+        try:
+            daemon = Daemon(*get_api_and_token(args.daemon_api, args.daemon_path))
+        except Exception  as exp:
```
Comment (Breigner01) [inline lotus-exporter-farcaster/lotus-exporter-farcaster.py:1230]: Do not except bare Exception

File: lotus-exporter-farcaster/lotus-exporter-farcaster.py [hunk 5]
```diff
@@ -1292,43 +1185,119 @@ def main():
...
+            raise DaemonError from exp
```
Comment (Breigner01) [inline lotus-exporter-farcaster/lotus-exporter-farcaster.py:1231]: Should change according to new name

File: lotus-exporter-farcaster/lotus-exporter-farcaster.py [hunk 6]
```diff
@@ -1292,43 +1185,119 @@ def main():
...
 
-    # Declare Global Variables
-    START_TIME = None
-    LOTUS_OBJ = None
-    METRICS_OBJ = None
+        try:
+            miner = Miner(*get_api_and_token(args.miner_api, args.miner_path))
+        except Exception  as exp:
```
Comment (Breigner01) [inline lotus-exporter-farcaster/lotus-exporter-farcaster.py:1235]: Do not except bare Exception

File: lotus-exporter-farcaster/lotus-exporter-farcaster.py [hunk 7]
```diff
@@ -1292,43 +1185,119 @@ def main():
...
+            raise MinerError from exp
```
Comment (Breigner01) [inline lotus-exporter-farcaster/lotus-exporter-farcaster.py:1236]: Should change according to new name

**Review by Breigner01 — CHANGES_REQUESTED** [review]
You've done a really good job overall, there are just some changes to follow the standard coding style.
I've also seen that you've removed all the try except on requests on api but I think that they should stay since it can fail and it should be handled properly.
We're now waiting for @s0nik42 to review the parts I asked him to.

**Pull request merged**